### ISSUE TITLE
[Feature] K3s cloud provider and csi driver support

### ISSUE THREAD

**Issue by yasker** [harvester/harvester#1812]
We need to support k3s cloud provider and CSI driver as well.

It needs to be GA quality with RKE2 driver for the next Rancher release (v2.6.4).

**Comment by FrankYang0529** [comment]
I can run harvester-csi-driver and create, read, write, resize and delete volume on k3s by following steps.

**Deploy harvester-csi-driver**
1. Create k3s cluster.
2. Run [this script](https://github.com/harvester/harvester-csi-driver/blob/master/deploy/generate_addon.sh) to create a new kube-config file for k3s.
3. Copy kube-config file to `/etc/kubernetes/cloud-config` k3s VM.
4. Install harvester-csi-driver chart without changing any values.
5. `harvester-csi-driver` should be deployed successfully.

![Screen Shot 2022-02-08 at 6 17 08 PM](https://user-images.githubusercontent.com/4927361/152967153-b36fab40-cc89-40e3-880b-4746742e2cc9.png)

**Create, read and write volume with harvester storage class**
1. Create a Nginx deployment with storage. We can see related volume is created in Harvester.
![Screen Shot 2022-02-08 at 6 36 06 PM](https://user-images.githubusercontent.com/4927361/152970374-052263e2-608c-4c5a-8cf7-90f121887c82.png)
![Screen Shot 2022-02-08 at 6 37 15 PM](https://user-images.githubusercontent.com/4927361/152970548-56087a0c-27a1-45ad-9488-0cace45d07ad.png)
![Screen Shot 2022-02-08 at 6 38 52 PM](https://user-images.githubusercontent.com/4927361/152970941-d968f6cc-cacb-49de-890d-685efa396420.png)

2. Click `Execute shell` to access Nginx container.
3. Run `dd if=/dev/zero of=/test/tmpfile bs=1M count=512`
4. Run `dd if=/dev/null of=/test/tmpfile bs=1M count=512`
![Screen Shot 2022-02-08 at 5 51 58 PM](https://user-images.githubusercontent.com/4927361/152970988-fced108f-3c6e-451d-a187-5871c127398e.png)
5. Delete the Nginx deployment.

**Resize and delete volume with harvester storage class**
1. Change to Harvester dashboard.
2. Click the `Edit config` for the volume.
3. Change volume size. We can see the volume is in `Resizing` status.
![Screen Shot 2022-02-08 at 5 55 39 PM](https://user-images.githubusercontent.com/4927361/152971292-aabcb160-2955-4e6e-b157-6a4b4a7f2e0a.png)
4. Delete PVC in k3s dashboard.
5. Related volume should be deleted in Harvester too.

I will research how to add `harvester-csi-driver` chart as an option to k3s.

**Comment by FrankYang0529** [comment]
I can run harvester-cloud-provider with k3s to create dhcp/pool LoadBaclander by following steps.

**Setup k3s cluster**
1. Create k3s `1.21.9+k3s1` cluster.
![Screen Shot 2022-02-10 at 5 01 13 PM](https://user-images.githubusercontent.com/4927361/153373297-18a4c780-d9a1-4fad-82c0-50a81ad1d6b5.png)
2. Set `cloud-provider=external` in kubelet args.
![Screen Shot 2022-02-10 at 5 03 23 PM](https://user-images.githubusercontent.com/4927361/153373558-a496d5c7-3a0d-4e44-86f3-fa8d68f27155.png)
3. After k3s is `Active`, click `Edit YAML`.
![Screen Shot 2022-02-10 at 5 08 44 PM](https://user-images.githubusercontent.com/4927361/153374379-ff16dff4-efaa-43c6-b43f-b29d0c9e65e1.png)
4. Add `disable-cloud-controller: true`, click `Save`, and wait for k3s `Active` again.
![Screen Shot 2022-02-10 at 5 09 19 PM](https://user-images.githubusercontent.com/4927361/153374667-04cd7463-ccca-4281-9b01-c4ade4841123.png)

**Deploy harvester-cloud-provider**
1. Run [this script](https://github.com/harvester/cloud-provider-harvester/blob/master/deploy/generate_addon.sh) to create a new kube-config file for k3s.
2. Copy kube-config file to /etc/kubernetes/cloud-config k3s VM.
3. Install harvester-cloud-provider chart. Set Cloud config file path as `/etc/kubernetes/cloud-config`. Set Cluster name as your cluster name.
![Screen Shot 2022-02-10 at 4 19 47 PM](https://user-images.githubusercontent.com/4927361/153376495-8a317bc9-d887-4bdd-bc4c-156e2f0ad580.png)
4. harvester-cloud-provider should be deployed successfully.
![Screen Shot 2022-02-10 at 5 22 21 PM](https://user-images.githubusercontent.com/4927361/153376793-2c51b212-e8a3-4945-bc8b-804be559abbe.png)

**Create Nginx workload for testing**
1. Create a `test-nginx` deployment with image `nginx:latest`.
2. Add pod label `test: test`.
![Screen Shot 2022-02-10 at 4 21 19 PM](https://user-images.githubusercontent.com/4927361/153377040-4d8032c2-1aa6-4205-b06b-573d04b81498.png)

**Create a DHCP LoadBalancer**
1. Open Kubectl shell.
2. Create `test-dhcp-lb.yaml` file.
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    cloudprovider.harvesterhci.io/ipam: dhcp
  name: test-dhcp-lb
  namespace: default
spec:
  ports:
  - name: http
    nodePort: 30172
    port: 8080
    protocol: TCP
    targetPort: 80
  selector:
    test: test
  sessionAffinity: None
  type: LoadBalancer
```
3. Apply it.
![Screen Shot 2022-02-10 at 4 37 57 PM](https://user-images.githubusercontent.com/4927361/153377344-e16721a9-e982-480b-bfc5-37933abe533c.png)
4. The DHCP LoadBalancer should be created and should work.
![Screenshot from 2022-02-10 16-34-58](https://user-images.githubusercontent.com/4927361/153377738-c80877e8-2515-4000-8975-aa83b1faddfe.png)
5. Delete the DHCP LoadBalancer and should not have any error.

**Create a Pool LoadBalancer**
1. Modify vip-pool in Harvester settings.
![Screen Shot 2022-02-10 at 5 29 21 PM](https://user-images.githubusercontent.com/4927361/153378149-1274dbc6-5843-4351-ac1c-995867b32a7c.png)
2. Open Kubectl shell.
3. Create `test-pool-lb.yaml` file.
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    cloudprovider.harvesterhci.io/ipam: pool
  name: test-pool-lb
  namespace: default
spec:
  ports:
  - name: http
    nodePort: 32155
    port: 8080
    protocol: TCP
    targetPort: 80
  selector:
    test: test
  sessionAffinity: None
  type: LoadBalancer
```
4. Apply it.
![Screen Shot 2022-02-10 at 4 58 12 PM](https://user-images.githubusercontent.com/4927361/153378305-2dc1ad2a-a77f-4d35-9c2d-d5a6d7de3dfe.png)
5. The Pool LoadBalancer should get an IP from vip-pool and work.
![Screenshot from 2022-02-10 16-58-44](https://user-images.githubusercontent.com/4927361/153378595-362c1fad-5390-4f33-a620-9c1bf6b31976.png)
6. Delete the Pool LoadBalancer and should not have any error.

**Comment by TachunLin** [comment]
### Test result of K3s cloud provider: `PASS`
Verified **fixed** on rancher `v2.6-head` (99974f0) with Harvester master-75cc3ced-head (03/14). 

1. Can deploy K3s cluster on harvester with kubernetes version `v1.21.10+k3s1`
  ![image](https://user-images.githubusercontent.com/29251855/158532075-c7211ae9-d199-4fa5-9e3b-9ae68c7db055.png)
2. Can correctly install harvester cloud provider on K3s cluster 
  ![image](https://user-images.githubusercontent.com/29251855/158512528-42ff575a-87a6-4424-bfb5-fa7af94ea74d.png)
3. Can create a `DHCP LoadBalancer` on K3s cluster with no error
  ![image](https://user-images.githubusercontent.com/29251855/158513800-70d7c0ba-5a4b-4462-90df-c8d05b5a389d.png)
4. Can create a `Pool LoadBalancer` on K3s cluster with no error
  ![image](https://user-images.githubusercontent.com/29251855/158514315-1b570f64-fe18-400e-acc4-56d03bc30e61.png)

### Test Information
* Test Environment: 3 nodes harvester on provo bare machine
* Harvester version: master-75cc3ced-head
* Rancher version: v2.6-head (99974f0)
* Kubernetes version: v1.21.10+k3s1

### Verify Steps 
Ref: [K3s cloud provider](https://docs.harvesterhci.io/v1.0/rancher/cloud-provider/#deploying-to-the-k3s-cluster-with-harvester-node-driver-experimental)
1. Create a K3s cluster, select `v1.21.10+k3s1`
  ![image](https://user-images.githubusercontent.com/29251855/158510521-deef3d05-881e-4655-8ffe-5e4b9755dcad.png)
2. Click the Edit yaml button
3. Set `disable-cloud-provider: true` to disable default k3s cloud provider.
  ![image](https://user-images.githubusercontent.com/29251855/158510820-4d8a0021-1675-4c92-86b9-a6427f2e382b.png)
4. Add `cloud-provider=external` to use harvester cloud provider.
  ![image](https://user-images.githubusercontent.com/29251855/158511002-47a4a532-7f67-4eb0-8da4-074c6d9752e9.png)
5. Create K3s cluster 
![image](https://user-images.githubusercontent.com/29251855/158511706-1c0c6af5-8909-4b1d-bc2a-0fa2fa26e000.png)
6. Download the [Generate addon configuration](https://github.com/harvester/cloud-provider-harvester/blob/master/deploy/generate_addon.sh) for cloud provider 
7. Download Harvester kubeconfig and add into your local ~/.kube/config file 
8. Generate K3s kubeconfig by running generate addon script 
  ` ./deploy/generate_addon.sh <k3s cluster name> <namespace>`
  e.g `./generate_addon.sh k3s-focal-cloud-provider default`
9. Copy the kubeconfig content 
10. ssh to K3s VM
  ![image](https://user-images.githubusercontent.com/29251855/158534901-8fd22159-6a04-4592-ba25-ba4d73742a20.png)
11. Add kubeconfig content to `/etc/kubernetes/cloud-config` file, remember to align the yaml layout
12. Install Harvester cloud provider 
  ![image](https://user-images.githubusercontent.com/29251855/158512528-42ff575a-87a6-4424-bfb5-fa7af94ea74d.png)
  ![image](https://user-images.githubusercontent.com/29251855/158512667-18b0249c-f859-4ae4-96b7-42ce873cb97a.png)
  ![image](https://user-images.githubusercontent.com/29251855/158512758-d06df2f6-7094-4d41-b960-d50b26cd23fb.png)
#### Create Nginx workload for testing
13. Create a test-nginx deployment with image nginx:latest.
  ![image](https://user-images.githubusercontent.com/29251855/158512919-a35a079a-aa75-4ce8-bac6-a79438a2e112.png)  
14. Add pod label test: test.
  ![image](https://user-images.githubusercontent.com/29251855/158513017-5afc909a-662a-4f4e-b867-2555241a2cbd.png)
  ![image](https://user-images.githubusercontent.com/29251855/158513105-09ab472b-7cd4-4352-b4e1-84f673ee7088.png)
#### Create a DHCP LoadBalancer
15. Open Kubectl shell.
16, Create `test-dhcp-lb.yaml` file.
  ```
  apiVersion: v1
  kind: Service
  metadata:
    annotations:
      cloudprovider.harvesterhci.io/ipam: dhcp
    name: test-dhcp-lb
    namespace: default
  spec:
    ports:
    - name: http
      nodePort: 30172
      port: 8080
      protocol: TCP
      targetPort: 80
    selector:
      test: test
    sessionAffinity: None
    type: LoadBalancer
  ```
17. Run `k apply -f test-dhcp-lb.yaml` to apply it.
  ![image](https://user-images.githubusercontent.com/29251855/158513659-3e0c487b-c819-492c-8c62-62fc644fd858.png)
18. The Pool LoadBalancer should get an IP from vip-pool and work.
  ![image](https://user-images.githubusercontent.com/29251855/158513800-70d7c0ba-5a4b-4462-90df-c8d05b5a389d.png)

#### Create a Pool LoadBalancer
19. Modify vip-pool in Harvester settings.
  ![image](https://user-images.githubusercontent.com/29251855/158514040-bfcd9ff3-964a-4511-94d7-a497ef88848f.png)

20. Open Kubectl shell.
21. Create `test-pool-lb.yaml` file.
  ```
  apiVersion: v1
  kind: Service
  metadata:
    annotations:
      cloudprovider.harvesterhci.io/ipam: pool
    name: test-pool-lb
    namespace: default
  spec:
    ports:
    - name: http
      nodePort: 32155
      port: 8080
      protocol: TCP
      targetPort: 80
    selector:
      test: test
    sessionAffinity: None
    type: LoadBalancer
  ```
22. Run `k apply -f test-pool-lb.yaml` to apply it.
23. The Pool LoadBalancer should get an IP from vip-pool and work.
![image](https://user-images.githubusercontent.com/29251855/158514315-1b570f64-fe18-400e-acc4-56d03bc30e61.png)

**Comment by TachunLin** [comment]
### Test result of K3s csi driver: `Ongoing`

**Comment by TachunLin** [comment]
### Test result of K3s csi driver:

Verified **fixed** on rancher `v2.6.4-rc10` with Harvester `v1.0-1-rc1`.

1. Can deploy K3s cluster on harvester with kubernetes version `v1.23.4+k3s1`
  ![image](https://user-images.githubusercontent.com/29251855/158935474-9d6f1c37-ea59-485d-83f5-6f1b19ebfa98.png)

2. Can correctly install harvester csi driver on K3s cluster
  ![image](https://user-images.githubusercontent.com/29251855/158933473-a0172a62-5f7c-4e68-860c-c9cb38275791.png)
  ![image](https://user-images.githubusercontent.com/29251855/158933756-d4198111-ae05-4a7d-8ea8-d21e3f3d2b87.png)

3. Can deploy nginx service with new PVC created
  ![image](https://user-images.githubusercontent.com/29251855/158934469-6b050d39-a45c-493d-9e29-89e16c1cf23d.png)
  ![image](https://user-images.githubusercontent.com/29251855/158934499-6c7d9525-f43a-4426-8786-1e4aec099964.png)
  ![image](https://user-images.githubusercontent.com/29251855/158934531-9c42b704-67cc-4c91-b24e-fa29f7183f05.png)
4. Can allocate size in nginx container 
  ![image](https://user-images.githubusercontent.com/29251855/158934986-c08ddccc-3b33-4508-9861-6d10c5ded3c5.png)
5. Can resize and delete volume with harvester storage class
  ![image](https://user-images.githubusercontent.com/29251855/158935263-a8b6fa9e-1f4b-43ae-a687-6df3e34986a1.png)
  ![image](https://user-images.githubusercontent.com/29251855/158935288-67409de3-d5d5-4c9d-af04-421f4153658b.png)

### Test Information
* Test Environment: 3 node harvester on provo bare machines
* Harvester version: v1.0.1-rc1
* Rancher version: v2.6.4-rc4
* Kubernetes version: v1.23.4+k3s1

### Verify Steps
Ref: [K3s node driver](https://docs.harvesterhci.io/v1.0/rancher/csi-driver/#deploying-with-harvester-k3s-node-driver)
#### Deploy harvester-csi-driver
1. Create a K3s cluster, select `v1.23.4+k3s1`
  ![image](https://user-images.githubusercontent.com/29251855/158931446-9bf36549-1f45-4ed0-931e-e0a09eb2f5e6.png)
  ![image](https://user-images.githubusercontent.com/29251855/158931345-96d1851f-d934-4b5f-b05e-9582e318401c.png)

2. Add the following user data
  ```
  password: 123456
  chpasswd: { expire: False }
  ssh_pwauth: True
  ```
3. Create K3s cluster
 ![image](https://user-images.githubusercontent.com/29251855/158932127-6bee7392-3b0a-4097-b739-d19b307e801d.png)

4. Download the [Generate addon configuration](https://github.com/harvester/harvester-csi-driver/blob/master/deploy/generate_addon.sh) for csi driver
6. ssh to harvester node 1
7. Execute `cat /etc/rancher/rke2/rke2.yaml`  
8. change the server value from https://127.0.0.1:6443 to your node1 IP
9. Copy the kubeconfig and add into your local ~/.kube/config file
10. Generate K3s kubeconfig by running generate addon script
` ./deploy/generate_addon.sh <k3s cluster name> <namespace>`
e.g `./generate_addon.sh k3s-csi-cluster default`
11. Copy the ks3 kubeconfig content start with `cloud-provider-config:` between `addons_include:`
12. ssh to K3s VM
  ![image](https://user-images.githubusercontent.com/29251855/158932912-38297b70-7546-4349-801f-6a4b8b973305.png)

13. Add kubeconfig content to `/etc/kubernetes/cloud-config`file, remember to align the yaml layout
14. Install Harvester csi driver
  ![image](https://user-images.githubusercontent.com/29251855/158550983-61cff655-66a6-4a49-96bd-6e208f4fc9d8.png)
  ![image](https://user-images.githubusercontent.com/29251855/158551034-090cee3d-9bd6-425c-84d8-16f6a66a5c64.png)
  ![image](https://user-images.githubusercontent.com/29251855/158933473-a0172a62-5f7c-4e68-860c-c9cb38275791.png)
  ![image](https://user-images.githubusercontent.com/29251855/158933756-d4198111-ae05-4a7d-8ea8-d21e3f3d2b87.png)

#### Create Nginx workload for testing
15. Create a nginx-csi deployment with image nginx:latest.
  ![image](https://user-images.githubusercontent.com/29251855/158934043-11d94957-5b85-469e-bdf7-5658edfec5d9.png)


16. Create a new PVC in storage tab: 
  ![image](https://user-images.githubusercontent.com/29251855/158934170-8cb212e7-c84c-48d4-a15d-b921a9e6a8fc.png)

17. Complete the nginx deployment, create related PV in Harvester volume
  ![image](https://user-images.githubusercontent.com/29251855/158934469-6b050d39-a45c-493d-9e29-89e16c1cf23d.png)
  ![image](https://user-images.githubusercontent.com/29251855/158934499-6c7d9525-f43a-4426-8786-1e4aec099964.png)
  ![image](https://user-images.githubusercontent.com/29251855/158934531-9c42b704-67cc-4c91-b24e-fa29f7183f05.png)


18. Click Execute shell to access Nginx container.
19. Run `dd if=/dev/zero of=/test/tmpfile bs=1M count=512`
20. Run `dd if=/dev/null of=/test/tmpfile bs=1M count=512`
 ![image](https://user-images.githubusercontent.com/29251855/158934986-c08ddccc-3b33-4508-9861-6d10c5ded3c5.png)

21. Delete the Nginx deployment.

#### Resize and delete volume with harvester storage class
22. Change to Harvester dashboard.
23. Click the Edit config for the volume.
24. Change volume size. We can see the volume is in Resizing status.
![image](https://user-images.githubusercontent.com/29251855/158935263-a8b6fa9e-1f4b-43ae-a687-6df3e34986a1.png)
![image](https://user-images.githubusercontent.com/29251855/158935288-67409de3-d5d5-4c9d-af04-421f4153658b.png)
25. Delete PVC in k3s dashboard.
26. Related volume should be deleted in Harvester too.
